stub/src/main/java/io/grpc/stub/BlockingClientCall.java

```diff
@@ -0,0 +1,415 @@
+   * not guarantee that the server gets the message.
+   * </p>
+   * <p>
+   * <b>WARNING:  </b>Doing only writes without reads can lead to deadlocks as a result of flow
+   * control.  Furthermore, the server closing the stream will only be identified after the last
+   * sent value is read.
+   * </p>
+   *
+   * @param request Message to send to the server
+   * @param timeout How long to wait before giving up.  Values &lt;= 0 are no wait
+   * @param unit A TimeUnit determining how to interpret the timeout parameter
+   * @return true if the request is sent to stream, false if skipped
+   * @throws TimeoutException if write does not become ready before the specified timeout expires
+   * @throws io.grpc.StatusRuntimeException If the stream has closed in an error state
+   */
+  public boolean write(ReqT request, long timeout, TimeUnit unit)
+      throws InterruptedException, TimeoutException {
+    executor.drain();
+
+    if (getClosedStatus() != null && !getClosedStatus().isOk()) {
+      throw getClosedStatus().asRuntimeException();
+    }
+
+    if (writeClosed) {
+      throw new IllegalStateException("Writes cannot be done after calling halfClose or cancel");
+    } else if (!isWriteLegal()) {
+      return false;
+    }
+
+    boolean writeDone = false;
+
+    long start = System.nanoTime();
+    long duration = unit.toNanos(timeout);
+
+    while (!writeDone && !writeClosed) {
+      if (call.isReady()) {
+        call.sendMessage(request);
+        writeDone = true;
+      } else if (getClosedStatus() == null) {
+        waitAndDrainExecutorOrTimeout(start, duration);
+      } else {
+        break;
+      }
+    }
+
+    if (writeDone || getClosedStatus() == null) {
+      return writeDone;
+    }
+
+    if (getClosedStatus().isOk()) {
+      return false;
+    } else {
+      throw getClosedStatus().asRuntimeException();
+    }
+  }
+
+  /**
+   * Cancel stream and stop any further writes.  Note that some reads that are in flight may still
+   * happen after the cancel.
+   *
+   * @param message if not {@code null}, will appear as the description of the CANCELLED status
+   * @param cause if not {@code null}, will appear as the cause of the CANCELLED status
+   */
+  public void cancel(String message, Throwable cause) {
+    if (message == null) {
+      message = "User requested a cancel";
+    }
+    call.cancel(message, cause);
+    writeClosed = true;
+    executor.add(NoOpRunnable.INSTANCE);
+  }
+
+  /**
+   * Indicate that no more writes will be done and the stream will be closed from the client side.
+   * <br>
+   * See {@link ClientCall#halfClose()}
+   */
+  public void halfClose() {
+    boolean origWriteClosed = writeClosed;
+    writeClosed = true;
+
+    if (!origWriteClosed && closedStatus == null) {
+      call.halfClose();
+    } else if (origWriteClosed) {
+      throw new IllegalStateException("halfClose cannot be called after stream terminated");
```

Comment:
Why do you think that we would need an AtomicInteger?

Reply:
Because I thought the API in this class should be thread safe - if user has one thread to call `halfClose` and the other thread to call `cancel`, then the code won't be able to handle the timing condition.
However, if thread safety is not a concern, ([ClientCall](https://github.com/grpc/grpc-java/blob/master/api/src/main/java/io/grpc/ClientCall.java#L40) itself is not thread safe as publicly documented), then there is no need for atomicInteger.

Also the `getClosedStatus()` seem can be removed. Because there is an issue: 

t1: halfclose is called
t2: stream is closed and closedStatus=ok
t3: check closedSatus(L271)

then halfClose in t1 is lost.

Reply:
If one thread calls halfClose and the other cancel then it is valid both for the halfClose to succeed or to throw an exception.  The exception isn't thrown because it is causing the system problems, but rather because it likely indicates that there is a problem in the user's logic which they should look into.  
If the client calls halfClose after the server sends close it won't cause a problem, it is just inefficient.  In the example above, it is fine for halfClose not to be sent to the server because the server is by design no longer listening since it sent its own close.